### PR TITLE
Azure: fix building final url from endpoint object

### LINKED ISSUE #62735: Failed to start using 24.3.2.23-lts with Azure Blob Storage disk
**Describe what's wrong**

When upgrading from `24.2.1.2248-stable` to `24.3.2.23-lts`, ClickHouse has been unable to start, crashing with the following stack trace:
```
2024.04.17 13:29:27.877133 [ 1343786 ] {} <Error> Application: Code: 695. DB::Exception: Load job 'load table system.query_views_log' failed:
std::exception. Code: 1001, type: Azure::Storage::StorageException, e.what() = 400 The requested URI does not represent any resource on the server.
The requested URI does not represent any resource on the server.
RequestId:ac73a5ad-301e-0003-31cb-90beb5000000
Time:2024-04-17T13:29:27.3341509Z
Request ID: ac73a5ad-301e-0003-31cb-90beb5000000, Stack trace (when copying this message, always include the lines below):


0. ./contrib/llvm-project/libcxx/include/__memory/compressed_pair.h:36: Azure::Storage::StorageException::CreateFromResponse(std::unique_ptr<Azure::Core::Http::RawResponse, std::default_delete<Azure::Core::Http::RawResponse>>) @ 0x000000001512da16
1. ./build_docker/./contrib/azure/sdk/storage/azure-storage-blobs/src/rest_client.cpp:7746: Azure::Storage::Blobs::_detail::BlockBlobClient::StageBlock(Azure::Core::Http::_internal::HttpPipeline&, Azure::Core::Url const&, Azure::Core::IO::BodyStream&, Azure::Storage::Blobs::_detail::BlockBlobClient::StageBlockBlobBlockOptions const&, Azure::Core::Context const&) @ 0x000000001511e654
2. ./build_docker/./contrib/azure/sdk/storage/azure-storage-blobs/src/block_blob_client.cpp:0: Azure::Storage::Blobs::BlockBlobClient::StageBlock(String const&, Azure::Core::IO::BodyStream&, Azure::Storage::Blobs::StageBlockOptions const&, Azure::Core::Context const&) const @ 0x00000000150fc3d7
3. ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::WriteBufferFromAzureBlobStorage::writePart()::$_0::operator()() const::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000fa5fcd2
4. ./contrib/llvm-project/libcxx/include/__functional/function.h:0: ? @ 0x000000000fa5d0ea
5. ./contrib/llvm-project/libcxx/include/__functional/function.h:818: ? @ 0x000000000fa5f9fc
6. ./build_docker/./src/Common/ThreadPoolTaskTracker.cpp:118: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::TaskTracker::add(std::function<void ()>&&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000000fa61cd6
7. ./contrib/llvm-project/libcxx/src/future.cpp:221: std::packaged_task<void ()>::operator()() @ 0x000000000f751c7e
8. ./contrib/llvm-project/libcxx/include/future:2047: std::future<void> std::__function::__policy_invoker<std::future<void> (std::function<void ()>&&, Priority)>::__call_impl<std::__function::__default_alloc_func<DB::TaskTracker::syncRunner()::$_0, std::future<void> (std::function<void ()>&&, Priority)>>(std::__function::__policy_storage const*, std::function<void ()>&&, Priority&&) (.llvm.13264709103245809207) @ 0x000000000fa61ac7
9. ./contrib/llvm-project/libcxx/include/__functional/function.h:0: ? @ 0x000000000fa60d6a
10. ./contrib/llvm-project/libcxx/include/__functional/function.h:818: ? @ 0x000000000fa5e840
11. ./src/IO/WriteBuffer.h:65: DB::WriteBufferFromFileDecorator::nextImpl() @ 0x0000000010364a8f
12. ./src/IO/WriteBuffer.h:65: DB::WriteBufferFromFileDecorator::preFinalize() @ 0x000000001036465d
13. ./build_docker/./src/IO/WriteBufferFromFileDecorator.cpp:26: DB::WriteBufferFromFileDecorator::finalizeImpl() @ 0x0000000010364558
14. ./contrib/llvm-project/libcxx/include/__functional/function.h:863: ? @ 0x0000000010398c89
15. DB::WriteBuffer::finalize() @ 0x0000000007668e08
16. ./build_docker/./src/Disks/IDisk.cpp:0: DB::IDisk::checkAccessImpl(String const&) @ 0x00000000103470ec
17. ./contrib/llvm-project/libcxx/include/string:1499: DB::IDisk::startup(std::shared_ptr<DB::Context const>, bool) @ 0x0000000010346885
18. ./contrib/llvm-project/libcxx/include/string:1499: std::shared_ptr<DB::IDisk> std::__function::__policy_invoker<std::shared_ptr<DB::IDisk> (String const&, Poco::Util::AbstractConfiguration const&, String const&, std::shared_ptr<DB::Context const>, std::map<String, std::shared_ptr<DB::IDisk>, std::less<String>, std::allocator<std::pair<String const, std::shared_ptr<DB::IDisk>>>> const&, bool, bool)>::__call_impl<std::__function::__default_alloc_func<DB::registerDiskObjectStorage(DB::DiskFactory&, bool)::$_0, std::shared_ptr<DB::IDisk> (String const&, Poco::Util::AbstractConfiguration const&, String const&, std::shared_ptr<DB::Context const>, std::map<String, std::shared_ptr<DB::IDisk>, std::less<String>, std::allocator<std::pair<String const, std::shared_ptr<DB::IDisk>>>> const&, bool, bool)>>(std::__function::__policy_storage const*, String const&, Poco::Util::AbstractConfiguration const&, String const&, std::shared_ptr<DB::Context const>&&, std::map<String, std::shared_ptr<DB::IDisk>, std::less<String>, std::allocator<std::pair<String const, std::shared_ptr<DB::IDisk>>>> const&, bool, bool) @ 0x0000000010399d56
19. ./contrib/llvm-project/libcxx/include/__functional/function.h:0: ? @ 0x000000001033c080
20. ./build_docker/./src/Disks/DiskSelector.cpp:52: DB::DiskSelector::initialize(Poco::Util::AbstractConfiguration const&, String const&, std::shared_ptr<DB::Context const>, std::function<bool (Poco::Util::AbstractConfiguration const&, String const&, String const&)>) @ 0x00000000103402c3
21. ./contrib/llvm-project/libcxx/include/__functional/function.h:818: ? @ 0x00000000104b1fdd
22. ./build_docker/./src/Interpreters/Context.cpp:4160: DB::Context::getStoragePolicySelector(std::lock_guard<std::mutex>&) const @ 0x00000000104933e2
23. ./build_docker/./src/Interpreters/Context.cpp:4093: DB::Context::getStoragePolicy(String const&) const @ 0x00000000104b25ab
24. ./contrib/llvm-project/libcxx/include/__utility/swap.h:37: DB::MergeTreeData::getStoragePolicy() const @ 0x0000000011d3936b
25. ./src/Storages/MergeTree/MergeTreeData.h:927: DB::MergeTreeData::initializeDirectoriesAndFormatVersion(String const&, bool, String const&, bool) @ 0x0000000011d2e4f9
26. ./build_docker/./src/Storages/StorageMergeTree.cpp:120: DB::StorageMergeTree::StorageMergeTree(DB::StorageID const&, String const&, DB::StorageInMemoryMetadata const&, DB::LoadingStrictnessLevel, std::shared_ptr<DB::Context>, String const&, DB::MergeTreeData::MergingParams const&, std::unique_ptr<DB::MergeTreeSettings, std::default_delete<DB::MergeTreeSettings>>) @ 0x00000000120f64de
27. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:1460: DB::create(DB::StorageFactory::Arguments const&) @ 0x00000000120f2238
28. ./build_docker/./src/Storages/StorageFactory.cpp:224: DB::StorageFactory::get(DB::ASTCreateQuery const&, String const&, std::shared_ptr<DB::Context>, std::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, DB::LoadingStrictnessLevel) const @ 0x0000000011703c61
29. ./contrib/llvm-project/libcxx/include/string:2067: DB::createTableFromAST(DB::ASTCreateQuery, String const&, String const&, std::shared_ptr<DB::Context>, DB::LoadingStrictnessLevel) @ 0x0000000010181674
30. ./contrib/llvm-project/libcxx/include/string:1499: DB::DatabaseOrdinary::loadTableFromMetadata(std::shared_ptr<DB::Context>, String const&, DB::QualifiedTableName const&, std::shared_ptr<DB::IAST> const&, DB::LoadingStrictnessLevel) @ 0x00000000101abb62
31. ./contrib/llvm-project/libcxx/include/__functional/function.h:717: ? @ 0x00000000101b3b33
```

The code that seems to be failing was introduced in #59929.

**Does it reproduce on the most recent release?**

It happens in the most recent release.

**Enable crash reporting**

I cannot do this currently.

**How to reproduce**

I don't have a reproducer currently. This looks specific to using Azure Blob Storage disks but I haven't been able to reproduce it locally (with Azurite) as of now.

### PR DESCRIPTION
**Pull request by danipozo**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix failure to start when storage account URL has trailing slash.

Fixes #62735

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit c75574a1d66b8e9795cbee1209c3a7b26439ea85 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62850/c75574a1d66b8e9795cbee1209c3a7b26439ea85/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by danipozo** [comment]
None of the CI failures seems related to this change:
- 03094_grouparraysorted_memory
- Hung check failed, possible deadlock found
- 02428_parameterized_view
- 02352_interactive_queries_from_file: timed out
- 00992_system_parts_race_condition_zookeeper_long: timed out

File: src/Disks/ObjectStorages/AzureBlobStorage/AzureBlobStorageAuth.h [hunk 1]
```diff
@@ -21,6 +21,8 @@ struct AzureBlobStorageEndpoint
     String getEndpoint()
     {
         String url = storage_account_url;
+        if (url.ends_with('/'))
+          url = url.substr(0, url.size() - 1);
```
Comment (kitaisreal) [inline src/Disks/ObjectStorages/AzureBlobStorage/AzureBlobStorageAuth.h:25]: Better to use `pop_back`.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Disks/ObjectStorages/AzureBlobStorage/AzureBlobStorageAuth.h (hunk 1 above)
Comment (danipozo) [reply src/Disks/ObjectStorages/AzureBlobStorage/AzureBlobStorageAuth.h]: Done, thanks

**Review by kitaisreal — APPROVED** [review]

**Review by Algunenano — APPROVED** [review]

**Pull request merged**